src/Logic.hs

```diff
@@ -634,54 +639,57 @@ unintegrateAfter pid state = case Pr.lookupPullRequest pid state of
           | pid <- Pr.filterPullRequestsBy shouldUpdate state
           ] state
   where
+  satisfiedCheck = contextSatisfiesChecks (Pr.mandatoryChecks state) context
+  getNewStatus new old = if new `supersedes` old then new else old
+  newStatusState oldStatus = case oldStatus of
+    Pr.AnyCheck old -> Pr.AnyCheck $ getNewStatus newStatus old
+    Pr.SpecificChecks checks -> case satisfiedCheck of
+      Just check -> Pr.SpecificChecks $ Map.insertWith getNewStatus check newStatus checks
+      -- Ignore status updates that aren't relevant to the mandatory checks
+      Nothing -> Pr.SpecificChecks checks
+
   shouldUpdate pr = case Pr.integrationStatus pr of
-    Integrated candidateSha oldStatus -> candidateSha == buildSha && newStatus `supersedes` oldStatus
-    _                                 -> False
-  setBuildStatus pr = pr
-                    { Pr.integrationStatus = Integrated buildSha newStatus
-                    , Pr.needsFeedback = case newStatus of
-                                         BuildStarted _ -> True
-                                         BuildFailed _  -> True
-                                         _              -> Pr.needsFeedback pr -- unchanged
-                    }
-
--- | Does the first build status supersedes the second?
---
--- * The same status does not supersede itself
---
--- * Statuses 'BuildSuceeded' and 'BuildFailed' are not
---   superseded by other statuses
---
--- * We ignore changes of just the URL.
---
--- This is used in 'handleBuildStatusChanged`.
---
--- This is needed because there may be two concurrent builds pointing to the
--- same commit hash depending on the CI you are using with GitHub.
--- This guarantees that 'BuildSucceeded' or 'BuildFailed' statuses
--- cannot be overridden by later builds of a branch pointing to the same hash.
-supersedes :: BuildStatus -> BuildStatus -> Bool
-newStatus `supersedes` oldStatus        | sameStatus newStatus oldStatus
-                                        = False -- url change, ignore
-_         `supersedes` (BuildFailed _)  = False -- final status
-_         `supersedes` BuildSucceeded   = False -- final status
-_         `supersedes` _                = True
-
--- | Compares if two build statuses are the same
---   while ignoring any URL arguments
-sameStatus :: BuildStatus -> BuildStatus -> Bool
-sameStatus BuildStarted{} BuildStarted{} = True
-sameStatus BuildFailed{}  BuildFailed{}  = True
-sameStatus status1        status2        = status1 == status2
+    Integrated candidateSha _ -> candidateSha == buildSha
+    _                         -> False
+
+  -- Ensure that the pull request is actually being superceded by the failure,
+  -- before unintegrating.
+  unintegratePullRequestIfNeeded pid
+    | Just pr <- Pr.lookupPullRequest pid state
+    , Integrated _ status <- Pr.integrationStatus pr
+    , newStatus `supersedes` latestStatus status
+    , BuildFailed _ <- newStatus = unintegrateAfter pid
```

Comment:
Wouldn't this also unintegrate a PR if a build failure comes in for a non-mandatory check?

I think we should first incorporate the update into the `OutstandingCheck`s, then see if that made the summarized outcome turn to `BuildFailed`.